keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1244 @@
+| Pod                                       | *status.podIP*                     | slice            | *status.podIPs*                | *Set to "" if field is omitted*                               |
+| Pod                                       | status.nominatedNodeName           | string           | status.nominatedNodeName       |                                                               |
+| Secret                                    | type                               | enum             | type                           | string(enum)                                                  |
+
+[^1]: This field selector is commented with: "This is a bug, but we need to support it for backward compatibility.".
+
+Note that:
+
+- Resources always provide a value for a selectable field. Notice that in the
+  above table, *all* field selectors for optional fields provide a fallback if the
+  field is omitted.
+- Field selectors only exist for strings, enums, integers and booleans.
+- Field selectors for name and namespace are provided automatically.
+- Field selectors that do not match the path name of the actual field do exist,
+  for historical reasons, but the best practice is for field selectors to match
+  the exact path name of the field.
+
+Also note that we do not currently publish any field selector information into
+OpenAPIv3, and we do not document in APIs or on the Kubernetes website which
+fields may be selected. Because of this, finding the field selectors available for built-in types, to construct the above table, required
+locating [`GetAttrs/*ToSelectableFields` functions in the Kubernetes source code](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/strategy.go#L340).
+
+### Plan
+
+Because `generic.StoreOptions` consumes selectors from built-in and CRDs in the
+same way, we would extend the `GetAttrs` of CRDs to offer fields for selection.
+https://github.com/jpbetz/kubernetes/tree/crd-object-filters contains a working
+prototype.
+
+We would leverage the fieldPath validation supported added for CRD Validation
+Rules to validate paths.
+
+We would leverage the jsonpath library in the same way as used by
+additionalPrinterColumns to retrieve values from custom resources using
+jsonpaths.
+
+The majority of the remaining work will be to update the API to add the
+field and to validate it.
+
+#### API
+
+```go
+type CustomResourceDefinitionVersion struct {
+  
+  // selectableFields specifies paths to fields that may be used as field selectors.
+  // A maximum of 8 selectable fields are allowed.
+	// See https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors
+	SelectableFields []SelectableField `json:"selectableFields,omitempty" protobuf:"bytes,9,rep,name=selectableFields"`
+}
+
+  // SelectableField specifies the JSON path of a field that may be used with field selectors.
+type SelectableField struct {
```

Comment:
Let's work out whether we see potential eventual value in supporting CEL here to do things like the podIP selector.  If so, this would be a discriminated union (JSONPath or CEL as the top level choice).

@liggitt

Reply:
+1. I'll bring this up during API review.  I'd like to leave the door open for this in the future.

Reply:
Thinking on it, we could likely get similar power by using a combination of MutationAdmissionPolicy to do the computation and a single jsonPath.  That would leave the storage indexing more practical as well.  I think I'm at a point where I see jsonPath as sufficient.

Reply:
Keeping cel evaluations to the write path seems like a better idea to me